packages/tasit-action/src/ConfigLoader.test.js

```diff
@@ -1,40 +1,230 @@
 import ProviderFactory from "./ProviderFactory";
 import ConfigLoader from "./ConfigLoader";
+import { ethers } from "ethers";
+const { providers } = ethers;
+const {
+  JsonRpcProvider,
+  FallbackProvider,
+  InfuraProvider,
+  EtherscanProvider,
+} = providers;
+
+const parseNetworkNameFromEthres = networkName => {
+  if (networkName === "unknown") return "other";
+  if (networkName === "homestead") return "mainnet";
+  return networkName;
+};
+
+const extractProviderConfig = async provider => {
+  await provider.ready;
+  const network = await provider.getNetwork();
+  let { name: networkName } = network;
+  const {
+    pollingInterval,
+    apiAccessToken: infuraApiKey,
+    apiKey: etherscanApiKey,
+    connection,
+    providers,
+    path,
+  } = provider;
+
+  let config = {
+    network: parseNetworkNameFromEthres(networkName),
+    pollingInterval,
+  };
+
+  if (provider instanceof EtherscanProvider) {
+    config = {
+      ...config,
+      provider: "etherscan",
+      etherscan: { apiKey: etherscanApiKey },
+    };
+  } else if (provider instanceof InfuraProvider) {
+    config = {
+      ...config,
+      provider: "infura",
+      infura: { apiKey: infuraApiKey },
+    };
+  } else if (provider instanceof JsonRpcProvider) {
+    const { url, user, password, allowInsecure } = connection;
+    const { protocol, hostname, port } = new URL(url);
+    const jsonRpc = {
+      url: `${protocol}//${hostname}`,
+      port: Number(port),
+      user,
+      password,
+      allowInsecure,
+    };
+    config = { ...config, provider: "jsonrpc", jsonRpc };
+  } else if (provider instanceof FallbackProvider) {
+    config = { ...config, provider: "fallback" };
+  }
+
+  return config;
+};
+
+const fulfillWithDefaults = config => {
+  let { provider } = config;
+  let { jsonRpc, provider: providerType, infura, etherscan } = provider;
+
+  if (providerType === "etherscan") {
+    if (!etherscan) etherscan = { apiKey: null };
+    else {
+      const { apiKey } = etherscan;
+      etherscan = { ...etherscan, apiKey };
+    }
+
+    provider = { ...provider, etherscan };
+  } else if (providerType === "infura") {
+    if (!infura) infura = { apiKey: null };
+    else {
+      const { apiKey } = infura;
+      infura = { ...infura, apiKey };
+    }
+
+    provider = { ...provider, infura };
+  } else if (providerType === "jsonrpc") {
+    const { user, password, allowInsecure } = jsonRpc;
+    jsonRpc = { ...jsonRpc, user, password };
+
+    if (allowInsecure === undefined)
+      jsonRpc = { ...jsonRpc, allowInsecure: false };
+
+    provider = { ...provider, jsonRpc };
+  }
+
+  config = { ...config, provider };
+  return config;
+};
+
+const checkConfig = async config => {
+  const userConfig = fulfillWithDefaults(config);
```

Comment:
Maybe a less vague word than fulfill? Or add what we're fulfilling to the variable name?

Reply:
Is `fulfillConfigWithDefaults` a better name?

Reply:
Yeah, I like that better